packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts

```diff
@@ -0,0 +1,404 @@
+        key: "1",
+        name: "Dev-Compact",
+        screenshots: [
+          {
+            path: "/images/demo-datasource/dev-compact.png",
+          },
+        ],
+      },
+      {
+        id: "0",
+        key: "2",
+        name: "Dev",
+        screenshots: [
+          {
+            path: "/images/demo-datasource/dev.png",
+          },
+        ],
+      },
+    ],
+    phases: [
+      {
+        dateStarted: EXPERIMENT_START_DATE,
+        name: "",
+        reason: "",
+        coverage: 1,
+        condition: "",
+        namespace: { enabled: false, name: "", range: [0, 1] },
+        variationWeights: [0.3334, 0.3333, 0.3333],
+      },
+    ],
+  },
+];
+
+/**
+ * END Constants for Demo Datasource
+ */
+
+// region POST /demo-datasource-project
+
+type CreateDemoDatasourceProjectRequest = AuthRequest;
+
+type CreateDemoDatasourceProjectResponse = {
+  status: 200;
+  project: ProjectInterface;
+};
+
+/**
+ * POST /demo-datasource-project
+ * Create a demo-datasource-project resource
+ * @param req
+ * @param res
+ */
+export const postDemoDatasourceProject = async (
+  req: CreateDemoDatasourceProjectRequest,
+  res: Response<
+    CreateDemoDatasourceProjectResponse | PrivateApiErrorResponse,
+    EventAuditUserForResponseLocals
+  >
+) => {
+  req.checkPermissions("manageProjects", "");
+  req.checkPermissions("createDatasources", "");
+  req.checkPermissions("createMetrics", "");
+  req.checkPermissions("createAnalyses", "");
+
+  const { org } = getOrgFromReq(req);
+
+  const demoProjId = getDemoDatasourceProjectIdForOrganization(org.id);
+  const existingDemoProject: ProjectInterface | null = await findProjectById(
+    demoProjId,
+    org.id
+  );
+
+  if (existingDemoProject) {
+    res.status(200).json({
+      status: 200,
+      project: existingDemoProject,
+    });
+    return;
+  }
+
+  try {
+    const project = await createProject(org.id, {
+      id: demoProjId,
+      name: "GrowthBook Demo Project",
+      description: "GrowthBook Demo Project",
+    });
+    const datasource = await createDataSource(
+      org.id,
+      "GrowthBook Demo Postgres Datasource",
+      DATASOURCE_TYPE,
+      DEMO_DATASOURCE_PARAMS,
+      DEMO_DATASOURCE_SETTINGS,
+      undefined,
+      "Datasource used for demoing GrowthBook Metrics, Experiments, and Datasource Connections",
+      [project.id]
+    );
+
+    // Create metrics
+    const metrics = await Promise.all(
+      DEMO_METRICS.map(async (m) => {
+        return createMetric({
+          ...m,
+          organization: org.id,
+          owner: ASSET_OWNER,
+          userIdColumns: { user_id: "user_id" },
+          userIdTypes: ["user_id"],
+          datasource: datasource.id,
+          projects: [project.id],
+          tags: DEMO_TAGS,
+        });
+      })
+    );
+
+    const ratioMetric = await createMetric({
+      ...DEMO_RATIO_METRIC,
+      denominator: metrics.find(
+        (x) => x.name === "Purchases - Number of Orders (72 hour window)"
```

Comment:
This seems very fragile

Reply:
I created a constant name that can be used in both places, and made sure the denominator existed before doing anything with it.